clients/include/norm.hpp

```diff
@@ -21,28 +28,260 @@
+{
+    return clanhe_(norm_type, uplo, n, A, lda, work);
+}
+
+inline float
+    xlanhe(char* norm_type, char* uplo, int* n, rocblas_double_complex* A, int* lda, double* work)
+{
+    return zlanhe_(norm_type, uplo, n, A, lda, work);
+}
+
+inline void xaxpy(
+    int* n, float* alpha, rocblas_float_complex* x, int* incx, rocblas_float_complex* y, int* incy)
+{
+    return caxpy_(n, alpha, x, incx, y, incy);
+}
+
+inline void xaxpy(int*                    n,
+                  double*                 alpha,
+                  rocblas_double_complex* x,
+                  int*                    incx,
+                  rocblas_double_complex* y,
+                  int*                    incy)
+{
+    return zaxpy_(n, alpha, x, incx, y, incy);
+}
+
+/* ============== Norm Check for General Matrix ============= */
+/*! \brief compare the norm error of two matrices hCPU & hGPU */
+template <typename T, typename std::enable_if<!is_complex<T>>::type* = nullptr>
+inline double norm_check_general(
+    char norm_type, rocblas_int M, rocblas_int N, rocblas_int lda, T* hCPU, T* hGPU)
+{
+    // norm type can be 'O', 'I', 'F', 'o', 'i', 'f' for one, infinity or Frobenius norm
+    // one norm is max column sum
+    // infinity norm is max row sum
+    // Frobenius is l2 norm of matrix entries
+
+    double error_double = std::numeric_limits<double>::quiet_NaN();
+
+    host_vector<double> hCPU_double(N * lda);
+    host_vector<double> hGPU_double(N * lda);
+
+    for(rocblas_int i = 0; i < N * lda; i++)
+    {
+        hCPU_double[i] = double(hCPU[i]);
+        hGPU_double[i] = double(hGPU[i]);
+    }
+
+    double      work[1];
+    rocblas_int incx  = 1;
+    double      alpha = -1.0;
+    rocblas_int size  = lda * N;
+
+    double cpu_norm = dlange_(&norm_type, &M, &N, hCPU_double.data(), &lda, work);
+    daxpy_(&size, &alpha, hCPU_double.data(), &incx, hGPU_double.data(), &incx);
+
+    double error = dlange_(&norm_type, &M, &N, hGPU_double.data(), &lda, work) / cpu_norm;
+
+    return error;
+}
+
+template <typename T, typename std::enable_if<is_complex<T>>::type* = nullptr>
+inline double norm_check_general(
+    char norm_type, rocblas_int M, rocblas_int N, rocblas_int lda, T* hCPU, T* hGPU)
+{
+    // norm type can be O', 'I', 'F', 'o', 'i', 'f' for one, infinity or Frobenius norm
+    // one norm is max column sum
+    // infinity norm is max row sum
+    // Frobenius is l2 norm of matrix entries
+
+    decltype(std::real(*hCPU)) work[1];
+    rocblas_int                incx  = 1;
+    decltype(std::real(*hCPU)) alpha = -1.0f;
+    rocblas_int                size  = lda * N;
+
+    double cpu_norm = xlang(&norm_type, &M, &N, hCPU, &lda, work);
+    xaxpy(&size, &alpha, hCPU, &incx, hGPU, &incx);
+
+    double error = xlang(&norm_type, &M, &N, hGPU, &lda, work) / cpu_norm;
+
+    return error;
+}
 
-/*! \brief  Template: norm check for hermitian/symmetric Matrix: half/float/double/complex */
+template <>
```

Comment:
TODO: Consolidate these functions into a smaller number of templates , varying them only by template arguments.

Reply:
Need to wrap rocblas_half to a class form to get conversion operation from half to double.
I think this can be done later.